### PR TITLE
Add ForwardSecurity and ForwardClient to support handshake

### PR DESCRIPTION
**Pull request by okkez**
These configurations are ported from Fluentd's in_forward.

See https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1

### CODE REVIEW

**Comment by okkez** [comment]
I'm working to support handshake phase in [Forward Protocol Specification v1](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1).
This PR is the first step to support handshake phase.

How do you think about this?

File: influent-java/src/main/java/influent/forward/ForwardClient.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      return this;
+    }
+
+    /**
+     * Set network address specification
+     *
+     * @param network Network address specification
+     * @return this builder
+     */
+    public Builder network(String network) {
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/ForwardClient.java:49]: What is difference between host and network?
I may not understand ip/network authentication……

**Comment by okumin** [comment]
It is good to implement [handshake protocol](https://github.com/okumin/influent/issues/9).
I wonder

* whether `shared_key` should be prepared per client hostname or not
    * I think that all the client can MITM when one key is shared by all the client and server and TLS is disabled
    * But Fluentd is using a single key per port
* how to manage client's passwords
    * I would not like to store plain passwords

File: influent-java/src/main/java/influent/forward/ForwardClient.java (hunk 1 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/ForwardClient.java:49]: https://github.com/fluent/fluentd/blob/v0.14.20/lib/fluent/plugin/in_forward.rb#L118-L142

Fluentd's in_forward treat that `host` and `network` are exclusive each other.
`host` means that client source IP address (or calculated from hostname). ex. `example.com` or `192.168.1.11`.
`network` means that client source IP address range. ex. `192.168.1.0/32`

For example, If `network = "192.168.1.0/24"`:
* client's IP address is `192.168.1.1`: OK
* client's IP address is `192.168.2.1`: NG (refused by server)

**Comment by okkez** [comment]
> whether shared_key should be prepared per client hostname or not 

Fluentd's in_forward supports one key shared by all clients for backward compatibility, I think.
Fluentd supports both per client shared key and global(shared by all clients) shared key.

See also
* https://github.com/fluent/fluentd/blob/v0.14.20/example/in_forward_client.conf
* https://github.com/fluent/fluentd/blob/v0.14.20/example/in_forward_shared_key.conf

> how to manage client's passwords 

is this influent scope? Influent users such as kafka-connect-fluentd should consider this.

**Comment by okumin** [comment]
> is this influent scope? Influent users such as kafka-connect-fluentd should consider this.

I have imagined a situation that hashed passwords are stored in a configuration file for security.
This is different from Fluentd since Fluentd loads raw passwords in configuration.

But such a situation never occur because a salt for password is generated per connection and hash with salt cannot be computed in advance.

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_forward.rb#L190-L191

So Influent should accept raw password, probably.

**Comment by okumin** [comment]
> Fluentd's in_forward supports one key shared by all clients for backward compatibility, I think.
> Fluentd supports both per client shared key and global(shared by all clients) shared key.

> Fluentd's in_forward treat that host and network are exclusive each other.
> host means that client source IP address (or calculated from hostname). ex. example.com or 192.168.1.11.
> network means that client source IP address range. ex. 192.168.1.0/32

I confirmed the implementation of ip/network authentication.
And I recognized shared_key per connection used in ip/network authentication(but I wonder why a shared key for user/password authentication is shared between all the client).
Do you know the location of ip/network authentication specification?

**Comment by okkez** [comment]
> Do you know the location of ip/network authentication specification?

See https://github.com/fluent/fluentd/blob/v0.14.20/lib/fluent/plugin/in_forward.rb#L415-L441

If `client.host` or `client.network` match remote address, in_forward find a node from `@nodes` constructed [here](https://github.com/fluent/fluentd/blob/v0.14.20/lib/fluent/plugin/in_forward.rb#L137).

If client node is not found, the node is treated as anonymous client.
Anonymous client is allowed when `@security.allow_anonymous_source` is `true`.

Use IP/network authentication with global shared key when the client have no users.

**Comment by okumin** [comment]
> If client.host or client.network match remote address, in_forward find a node from @nodes constructed here.
>
> If client node is not found, the node is treated as anonymous client.
> Anonymous client is allowed when @security.allow_anonymous_source is true.

I have understood that ip/network authentication's spec is only found on in_forward implementation.

**Comment by okkez** [comment]
#32 includes this PR.

**Pull request closed without merging**